### ISSUE TITLE
fullPage.js DOM structure is incorrect

### ISSUE THREAD

**Issue by alvarotrigo** [alvarotrigo/react-fullpage#33]
This issue is directly related with the reported bug trying to use the Parallax extension https://github.com/alvarotrigo/react-fullpage/issues/31

This is due to the wrong generation of the structure on the DOM:

![screen shot 2018-09-21 at 13 07 27](https://user-images.githubusercontent.com/1706326/45877808-5b0f9e00-bd9f-11e8-946f-06478989f774.png)

Notice how the direct child of the element with the class `fullpage-wrapper` is containing the anchor information in the `data-anchor` attribute and also other related ones like `data-fp-styles` which belong to a section.
```html
<div data-fp-styles="null" data-anchor="firstPage" style="height: 352px; background-color: rgb(40, 44, 52);">
    <div class="section section1 fp-section active fp-table fp-completely" data-fp-styles="null" data-anchor="firstPage" style="height: 352px; background-color: rgb(40, 44, 52);">
```

The wrapper around the sections (required by react) should not contain any attributes, like in [this jsfiddle](https://jsfiddle.net/ea17skjr/26/).

```html
<div>
    <div class="section section1 fp-section active fp-table fp-completely" data-fp-styles="null" data-anchor="firstPage" style="height: 352px; background-color: rgb(40, 44, 52);">
```

I'm not quite sure what's causing fullPage.js to add styles to that element and treating it like a section.

**Comment by alvarotrigo** [comment]
This might be the cause.
At the moment fullPage.js is initialised, this is the structure on the DOM:

![screen shot 2018-09-21 at 14 03 24](https://user-images.githubusercontent.com/1706326/45880108-24d61c80-bda7-11e8-9781-a34cd8db4963.png)

```html
<div id="react-root">
    <div id="fullpage">
        <div class="section"></div>
    </div>
</div>
```

We should only initialise fullPage.js once the whole structure is ready. 
Otherwise, we should be calling `componentDidUpdate` function when it is.

**Comment by alvarotrigo** [comment]
Apparently, fullPage.js is being initialised twice. 
The first one with the markup I mentioned on my previous answer.

Then, with this other one:
![screen shot 2018-09-21 at 14 10 08](https://user-images.githubusercontent.com/1706326/45880438-30761300-bda8-11e8-82b5-df328c912003.png)

**Comment by alvarotrigo** [comment]
Apparently removing a condition solves the issue.

I changed on the react component:

```javascript
render() {
    return (
      <div id="fullpage">
        {this.state.initialized
          ? this.props.render(this)
          : <div className="section" />
        }
      </div>
    );
  }
```

For this:
```javascript
render() {
    return (
      <div id="fullpage">
        {this.props.render(this) }
      </div>
    );
  }
```

Not quite sure the possible implications of it.

**Comment by cmswalker** [comment]
It should only mark as initialized once, but it will render twice. Once to establish the wrapper div with one section (required by fullpage.js js event hooks) and again when it gets the render prop HTML. I had to add the dummy section to resolve https://github.com/alvarotrigo/react-fullpage/issues/27 i can try it without to be sure

**Comment by cmswalker** [comment]
@alvarotrigo seems to be ok w/ the proposed change, thinking more about this, we should maybe just obfuscate the `initialized` state as it is really no value to the user any longer. Maybe state.__init or something. Basically I made it to where it would only render their provided HTML when initialized but we could just render regardless. Make sense?

**Comment by alvarotrigo** [comment]
Mmm ok, I'll give it another check!

**Comment by alvarotrigo** [comment]
> I'm not sure I understand why the dummy section is causing issues

The problem comes directly from the React script. I have no idea how it adds the new sections/markup on the existent element, but it does it the wrong way. It seems to be appending them as children of the existent dummy section instead of replacing the dummy section with the new content. 

Right after the first initialisation with the dummy section (which works as expected), React pushes the new content into the component and ends up creating this structure, which is just wrong:

![screen shot 2018-09-28 at 16 16 54](https://user-images.githubusercontent.com/1706326/46213864-fa9dd500-c339-11e8-9a70-0d0501aefbaa.png)

Any way we can fix this and completely remove the dummy section before appending the new content?

**Comment by alvarotrigo** [comment]
It also, somehow, removes the `fp-section` and `section` class from the dummy section **even before fullpage.js destroys itself** inside the `destroy` method. Making it impossible for fullPage.js to remove things like the `style` attribute.

**Comment by cmswalker** [comment]
It may just be a matter of transcluding the render prop differently. The dummy section was necessary to initialize fullpage.js with at least 1 section, otherwise there are runtime errors in fullpage.js as it cannot find any `.section` selectors. Once it's initialized w/ the dummy section we call render for the first time with the render prop and provided html. Can we do a side by side of the current markup vs the expected markup and I can try to make the change!

**Comment by cmswalker** [comment]
ok will look. What is the purpose of the `tableCell` class and why don't we want it in the expected markup after it's applied to the dummy section child?

**Comment by alvarotrigo** [comment]
> What is the purpose of the `tableCell` class and why don't we want it in the expected markup after it's applied to the dummy section child?

We basically need to replicate the required structure to use fullPage.js. `.fp-tablleCell` is an internal class and wrapper that fullPage.js adds in certain occasions (vertical alignment) that should have been removed when calling the destroy function of fullPage.js (which doesn't work as expected because React somehow removes the classes `fp-section` and `section` of the dummy section).

What it does is not very relevant for this issue. We just need to get rid of it as it is not part of the initial markup we pass on the react component. It just should't be there on the time of the initialisation or fullPage.js might not work as expected.

**Comment by alvarotrigo** [comment]
Btw, if solving this issue is too much of a hassle, perhaps we should consider if using a dummy section is the right solution for #27 ?
Seems kind of a hack to me ?

**Comment by cmswalker** [comment]
Ahh so I think the issue is that we aren't enforcing the use of [Fragments](https://reactjs.org/docs/fragments.html) This is the only way to get true transclusion with react components. As you can see in the example file the html we're handing down is wrapped in a dif with an id of `fullpage-wrapper` and that is weirding things out.

![screen shot 2018-10-04 at 7 35 59 pm](https://user-images.githubusercontent.com/8467234/46513373-0e9b7800-c80d-11e8-9e9c-8e0c3dcaf0b0.png)

However if you wrap in a fragment, things look better

![screen shot 2018-10-04 at 7 38 49 pm](https://user-images.githubusercontent.com/8467234/46513428-2f63cd80-c80d-11e8-9d2c-d3c2bc5453c0.png)

I can submit a PR and update the docs if this looks better. Seems to work fine, the dummy section is still needed to avoid a typeError in fullpage.js where it can't find any sections, so we have to render at least one. The lifecycle order goes

1. Render react component with the dummy section
2. Init fullpage.js code, it picks up the dummy section and all is ok
3. Component is now mounted, we can assign `fullpage_api` so the consumer has access to it
4. After the very first mount, re-render with provided markup from the user (render-prop) and pass the newly defined variables to them

**Comment by alvarotrigo** [comment]
> I can submit a PR and update the docs if this looks better.

Yeah that seems to solve the issue! 👍 

>  Seems to work fine, the dummy section is still needed to avoid a typeError in fullpage.js where it can't find any sections, so we have to render at least one. 

I'm not quite sure why it is necessary, but if there's anything I can do in fullpage.js itself that can help you deal with it, just let me know and I'll take a look. Perhaps we can avoid this step and do things in a less hacky way. 
Is it the fact that you need to be able to access the API even when there's no fullPage.js structure ready?

**Comment by cmswalker** [comment]
> Is it the fact that you need to be able to access the API even when there's no fullPage.js structure ready?

Yep, I believe that's it. Also because fullpage.js. I'll add a PR later today to address the layout issue!

**Comment by cmswalker** [comment]
fixed in @fullpage/react-fullpage@0.1.1
